File: python-test/pages/sauce_demo/shopping_cart_page.py
from typing import NamedTuple

from base_page import BasePage
from pages.sauce_demo.header_page import HeaderPage
from pages.sauce_demo.shopping_cart_elements import ShoppingCartElements

class ShoppingCartItem(NamedTuple):
    index: int
    quantity: int
    name: str
    price: str
    remove_from_cart: callable

class ShoppingCartPage(BasePage):

    def __init__(self, browser):
        super().__init__(browser)
        self.header = HeaderPage(browser)
        self.shopping_cart_elements = ShoppingCartElements(self.browser)
        self.set_browser(browser)

    def continue_shopping(self) -> None:
        self.shopping_cart_elements.continue_shopping_button.click()

    def get_cart_items(self) -> list[ShoppingCartItem]:
        items: list[ShoppingCartItem] = []
        item_entries = self.shopping_cart_elements.items_list()
        for index, _ in enumerate(item_entries):
            items.append(
                ShoppingCartItem(
                    index,
                    quantity = int(self.shopping_cart_elements.items_quantity()[index].text_content()),
                    name = self.shopping_cart_elements.items_name()[index].text_content(),
                    price = self.shopping_cart_elements.items_price()[index].text_content(),
                    remove_from_cart = lambda index=index: self.shopping_cart_elements.items_remove_from_cart.click(index)
                )
            )
        return items

File: typescript-test-playwright/pages/sauce-demo/shopping-cart.page.ts
import { Page } from '@playwright/test';
import { Header } from './header';

export type CartItems = {
  index: number;
  quantity: number;
  name: string;
  price: string;
  removeFromCart(): Promise<void>;
};

export class ShoppingCartPage {
  private page: Page;
  private _header: Header;

  constructor(page: Page) {
    this.page = page;
    this._header = new Header(this.page);
  }

  get header() {
    return this._header;
  }

  private readonly continueShoppingButton = '#continue-shopping';
  private readonly itemsList = '.cart_item';
  private readonly itemsQuantity = '.cart_quantity';
  private readonly itemsName = '.inventory_item_name';
  private readonly itemsPrice = '.inventory_item_price';
  private readonly itemsRemoveFromCart = 'button[id*="remove"]';

  async continueShopping() {
    await this.page.locator(this.continueShoppingButton).click();
  }

  async getCartItems(): Promise<CartItems[]> {
    const items: CartItems[] = [];
    const itemEntries = await this.page.locator(this.itemsList).all();
    for (const [index, item] of itemEntries.entries()) {
      items.push({
        index,
        quantity: parseInt(
          await item
            .locator(this.itemsQuantity)
            .textContent()
            .then((text) => text?.trim() ?? ''),
        ),
        name: await item
          .locator(this.itemsName)
          .textContent()
          .then((text) => text?.trim() ?? ''),
        price: await item
          .locator(this.itemsPrice)
          .textContent()
          .then((text) => text?.trim() ?? ''),
        removeFromCart: async () => {
          await item.locator(this.itemsRemoveFromCart).click();
        },
      });
    }
    return items;
  }
}
